File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int res = 0;
        int l = 0;
        int maxCharCount = 0;
        Map<Character, Integer> map = new HashMap<>();
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            map.put(c, map.getOrDefault(c, 0) + 1);
            maxCharCount = Math.max(maxCharCount, map.get(c));
            
            if (i - l + 1 - maxCharCount <= k)
                res = Math.max(res, i - l + 1);
            else {
                map.put(s.charAt(l), map.get(s.charAt(l)) - 1);
                l++;
            }
        }
        
        return res;
    }
}

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        res = 1
        l = 0
        r = 1
        
        hmap = {}
        hmap[s[0]] = 1
        
        while r < len(s):
            if s[r] not in hmap:
                hmap[s[r]] = 1
            else:
                hmap[s[r]] += 1
            max_count = max(hmap.values())
            length = r - l + 1
            need_replace = length - max_count
            
            if need_replace <= k:
                res = max(res, length)
                r += 1
            else:
                hmap[s[l]] -= 1
                l += 1
                r += 1
        
        return res